### PR TITLE
Added retry attempt when youtube-dl fails

### PR DESCRIPTION
**Pull request by kilmanio**
**Summary**
This PR fixes/implements the following **bugs/features**

* Ads a retry for youtube-dl is unable to extract video data

### CODE REVIEW

File: src/youtube_dl.rs [hunk 1]
```diff
@@ -28,21 +28,30 @@ where
...
     cmd.args(&ytdl_args);
     cmd.stdin(Stdio::null());
 
-    debug!("yt-dl command: {:?}", cmd);
-
-    let ytdl_output = cmd.output().await.unwrap();
-
-    if !ytdl_output.status.success() {
-        return Err(String::from_utf8(ytdl_output.stderr).unwrap());
+    let mut ytdl_output = cmd.output().await.unwrap();
```
Comment (Jokler) [inline src/youtube_dl.rs]: This makes it one more attempt than `max_attempts`.

File: src/youtube_dl.rs [hunk 2]
```diff
@@ -28,21 +28,30 @@ where
     Ok(dur.map(Duration::from_secs_f64))
 }
 
-pub async fn get_audio_download_url(uri: String) -> Result<AudioMetadata, String> {
+pub async fn get_audio_download_url(uri: String) -> Result<AudioMetadata, String>
```
Comment (Jokler) [inline src/youtube_dl.rs]: Missing `{` here.

File: src/youtube_dl.rs [hunk 3]
```diff
@@ -28,23 +28,39 @@ where
...
-        return Err(String::from_utf8(ytdl_output.stderr).unwrap());
+        let s = String::from_utf8(ytdl_output.stderr).unwrap();
+        return Err(s);
     }
 
     let output_str = String::from_utf8(ytdl_output.stdout).unwrap();
-    let output = serde_json::from_str(&output_str).map_err(|e| e.to_string())?;
+
+    Ok(output_str)
+}
```
Comment (fkaa) [inline src/youtube_dl.rs]: Move below `get_audio_download_url`

File: src/youtube_dl.rs [hunk 4]
```diff
@@ -28,23 +28,39 @@ where
...
+
+pub async fn get_audio_download_url(uri: String) -> Result<AudioMetadata, String> {
```
Comment (fkaa) [inline src/youtube_dl.rs]:
```suggestion
pub async fn get_audio_metadata_from_url(uri: String) -> Result<AudioMetadata, String> {
```

File: src/youtube_dl.rs [hunk 5]
```diff
@@ -28,23 +28,39 @@ where
...
-pub async fn get_audio_download_url(uri: String) -> Result<AudioMetadata, String> {
-    let ytdl_args = ["--no-playlist", "-f", "bestaudio/best", "-j", &uri];
+async fn run_youtube_dl(url: &str) -> Result<String, String> {
+    let ytdl_args = ["--no-playlist", "-f", "bestaudio/best", "-j", &url];
 
     let mut cmd = Command::new("youtube-dl");
     cmd.args(&ytdl_args);
     cmd.stdin(Stdio::null());
-
+
```
Comment (fkaa) [inline src/youtube_dl.rs]:
```suggestion

```

File: src/youtube_dl.rs [hunk 6]
```diff
@@ -28,23 +28,39 @@ where
...
+                run_youtube_dl(&uri).await?
+            } else {
+                return Err(e)
+            }
+        }
+    };
+
+    let output = serde_json::from_str(&ytdl_output).map_err(|e| e.to_string())?;
 
     Ok(output)
```
Comment (fkaa) [inline src/youtube_dl.rs]: Maybe something like this?

```suggestion
    match run_youtube_dl(&uri).await {
        Ok(o) => Ok(serde_json::from_str(&o).map_err(|e| e.to_string())?),
        Err(e) => {
            if e.contains("Unable to extract video data") {
                run_youtube_dl(&uri).await?
            } else {
                Err(e)
            }
        }
    }
```

**Review by fkaa — CHANGES_REQUESTED** [review]

**Pull request merged**